apis/database/v1alpha2/cloudsql_instance_types.go

```diff
@@ -53,84 +56,594 @@ const (
-	StorageGB int64 `json:"storageGB"`
+	// CreateTime: The time when the certificate was created in RFC 3339
+	// format, for example 2012-11-15T16:19:00.094Z
+	// +optional
+	CreateTime *string `json:"createTime,omitempty"`
 
-	// Tier (or machine type) for this instance, for example db-n1-standard-1
-	// (MySQL instances) or db-custom-1-3840 (PostgreSQL instances). For MySQL
-	// instances, this property determines whether the instance is First or
-	// Second Generation. For more information, see
-	// https://cloud.google.com/sql/docs/mysql/instance-settings
-	Tier string `json:"tier"`
+	// ExpirationTime: The time when the certificate expires in RFC 3339
+	// format, for example 2012-11-15T16:19:00.094Z.
+	// +optional
+	ExpirationTime *string `json:"expirationTime,omitempty"`
+
+	// Instance: Name of the database instance.
+	// +optional
+	Instance *string `json:"instance,omitempty"`
+
+	// SelfLink: The URI of this resource.
+	// +optional
+	SelfLink *string `json:"selfLink,omitempty"`
+
+	// Sha1Fingerprint: Sha1 Fingerprint.
+	// +optional
+	Sha1Fingerprint *string `json:"sha1Fingerprint,omitempty"`
+}
+
+// ReplicaConfiguration is Read-replica configuration for connecting to
+// the master.
+type ReplicaConfiguration struct {
+	// FailoverTarget is Specifies if the replica is the failover target. If
+	// the field is set to true the replica will be designated as a failover
+	// replica. In case the master instance fails, the replica instance will
+	// be promoted as the new master instance.
+	// Only one replica can be specified as failover target, and the replica
+	// has to be in different zone with the master instance.
+	// +optional
+	FailoverTarget *bool `json:"failoverTarget,omitempty"`
+
+	// MySQLReplicaConfiguration is MySQL specific configuration when
+	// replicating from a MySQL on-premises master. Replication
+	// configuration information such as the username, password,
+	// certificates, and keys are not stored in the instance metadata. The
+	// configuration information is used only to set up the replication
+	// connection and is stored by MySQL in a file named master.info in the
+	// data directory.
+	// +optional
+	MySQLReplicaConfiguration *MySQLReplicaConfiguration `json:"mysqlReplicaConfiguration,omitempty"`
+}
+
+// MySQLReplicaConfiguration is read-replica configuration specific to
+// MySQL databases.
+type MySQLReplicaConfiguration struct {
+	// CaCertificate: PEM representation of the trusted CA's x509
+	// certificate.
+	// +optional
+	CaCertificate *string `json:"caCertificate,omitempty"`
+
+	// ClientCertificate: PEM representation of the slave's x509
+	// certificate.
+	// +optional
+	ClientCertificate *string `json:"clientCertificate,omitempty"`
+
+	// ClientKey: PEM representation of the slave's private key. The
+	// corresponding public key is encoded in the client's certificate.
+	// +optional
+	ClientKey *string `json:"clientKey,omitempty"`
```

Comment:
This one seems complicated - as best as I can tell it's specifically configuring what private key to use for a MySQL read replica to connect to its master? Easy options seem to be:

1. Omit it, which I believe would mean we don't support configuring read replicas of SSL enabled MySQL instances.
1. Leave it as is, which means we'd support the above, but insecurely.

The other option would be to support reading it from a secret, but I'd rather avoid going down that path until we've had more time to think about the implications.

I'm inclined to go for option 2 in this case, then address this at some point after v1beta1 of this resource.

Reply:
> then address this at some point after v1beta1 of this resource.

I think after v1beta1 mark, it's easier to add fields but a lot harder to remove them. So, I'm inclined to keeping all the sensitive output fields in the secret. But if we really don't want to decide on its shape, removing the field now and not supporting SSL makes more sense.

Reply:
>  it's easier to add fields but a lot harder to remove them

That's a good point.

> I'm inclined to keeping all the sensitive output fields in the secret

Agreed, if this is in fact an output only field. It seemed to me like this might be an input field?

Reply:
There is `ServerCaCert *SslCert` only in `CloudsqlInstanceObservation`, so, it's definitely output-only.

But there are certificate-related fields also in `CloudsqlInstanceParameters.ReplicaConfiguration.MySQLReplicaConfiguration` and these are input parameters. Though, I am not fully sure about what we should do about them. According to GCP comment:
```
// MySQLReplicaConfiguration is MySQL specific configuration when
// replicating from a MySQL on-premises master. Replication
// configuration information such as the username, password,
// certificates, and keys are not stored in the instance metadata. The
// configuration information is used only to set up the replication
// connection and is stored by MySQL in a file named master.info in the
// data directory.
```

So, they are used only in the case that you manage your on-premise master and only get some replicas from CloudSQL, not sure how widely this feature is used. Requiring the user to supply a secret and having its ref here seems to be the safest way but I don't feel like we should do this now. Removing the SSL-related fields from `MySQLReplicaConfiguration` is another choice that's appealing. What do you think?

Reply:
> What do you think?

Yeah I pretty much exactly agree with what you said. i.e.

* Now is not the time to introduce the capability to _read_ sensitive inputs from a secret when creating a managed resource.
* I imagine this feature (TLS enabled MySQL read replicas) is niche enough that it would not be too impactful to lack support for it.
* It's easier to add a field (support for supplying this cert) later than it is to remove one.

@prasek any opinions here from a product perspective? The easy options are either:

* Don't support TLS enabled MySQL read replicas at the moment, because we don't have a secure way to configure the private key needed to do so. This is the more forward-compatible option.
* Support TLS enabled MySQL read replicas today, but do it insecurely by requiring you to specify a private key inline in your resource class. This is less forward-compatible because we'd likely remove this field once we figured out a better way.

Reply:
@negz inline private keys is probably a no-go, is that something the could be added with a secret ref with low lift vs. just not doing it?

Reply:
@prasek No - doing so has implications that I'd rather not block this work on.

Reply:
also [kubernetes api versioning](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning) gives us a fair bit of leeway here for v1beta1:
> The schema and/or semantics of objects may change in incompatible ways in a subsequent beta or stable release. When this happens, we will provide instructions for migrating to the next version. This may require deleting, editing, and re-creating API objects. The editing process may require some thought. This may require downtime for applications that rely on the feature.

Reply:
Here is what I understood from our discussion and what the latest commit does:
* Publish server's SSL certificate in the connection secret in case anyone wants to access it via SSL. Note that this cannot be supplied by the user, it's auto-generated by GCP. Though, you need to rotate it when its expiration is due, which is not an action we currently support. Details here https://cloud.google.com/sql/docs/mysql/configure-ssl-instance
* Read-replica CloudSQL for external master: SSL certificate configuration input is not supported since we don't yet have a good mechanism of having users provide input for a resource in the form of a secret (except provider). In fact, even if we supported it, there are some external manual configuration steps needed, see https://cloud.google.com/sql/docs/mysql/configure-ssl-instance

Reply:
Works for me, should we raise follow-on issues for (1) key-rotation and (2) read-replica private keys?

Reply:
@prasek FYI, I've opened the following issues to track these items.

(1) https://github.com/crossplaneio/stack-gcp/issues/53
(2) https://github.com/crossplaneio/stack-gcp/issues/54